pandas/core/apply.py

```diff
@@ -274,13 +274,16 @@ def apply_standard(self):
 
         # we cannot reduce using non-numpy dtypes,
         # as demonstrated in gh-12244
-        if (
+        can_reduce = (
             self.result_type in ["reduce", None]
             and not self.dtypes.apply(is_extension_array_dtype).any()
             # Disallow complex_internals since libreduction shortcut
             #  cannot handle MultiIndex
             and not isinstance(self.agg_axis, ABCMultiIndex)
-        ):
+        )
+        return_result = None
+
+        if can_reduce:
 
             values = self.values
```

Comment:
@Reksbril the underlying problem here is that this call to `self.values` is casting to object dtype which you want to avoid.  if you add to the check in 277-283 `and self.obj._is_homogeneous_type` that should solve the whole thing